pkg/asset/agent/manifests/nmstateconfig.go

```diff
@@ -356,16 +358,29 @@ func checkHostCount(installConfig *types.InstallConfig, agentConfig *agentconfig
 			// If not defined, the roles will be matched to replicas
 			if numMasters < numRequiredMasters {
 				numMasters++
-			} else if numWorkers < numRequiredWorkers {
+			} else {
 				numWorkers++
 			}
 		}
 	}
 
-	if numMasters != numRequiredMasters {
-		logrus.Warnf("The number of hosts configured as masters (%d) does not match the master replicas (%d)", numMasters, numRequiredMasters)
+	if numMasters != 0 {
+		if numMasters < numRequiredMasters {
+			return fmt.Errorf("not enough hosts found (%v) to support all the configured ControlPlane replicas (%v)", numMasters, numRequiredMasters)
+		}
+		if numMasters > numRequiredMasters {
```

Comment:
Technically this can't be true when numMasters == 0, since numRequiredMasters can't be negative so it doesn't matter, but it's a bit confusing having this inside the `if numMasters != 0` block.